Simpler & Scala.js-friendly `ClassTag.unapply`

adriaanm opened the pull request:
Follow up to #4235
- use `j.l.Class.isInstance` for Scala.js, so that `unapply`
  works correctly when referencing raw JavaScript classes (patch by @sjrd)
- inline overloaded calls to `unapply`, so we can get rid of them in 2.13

Review by @retronym, @som-snytt (lunch break permitting)

sjrd committed:
Scala.js-friendly `ClassTag.unapply`

Use `j.l.Class.isInstance` for Scala.js, so that `unapply`
works correctly when referencing raw JavaScript classes.

adriaanm committed:
Simplify `ClassTag.unapply`

Inline overloaded calls to `unapply`, so we can get rid of them in 2.13.

Note that theres a lot of `box(unbox(x))` going on behind the scenes.
Is this needed?

som-snytt commented:
The building had a power outage, so I'm on extended lunch break.

Per my comment on the previous PR, I'd recommend not shadowing `x` in the case statements, because that's where the extra boxing comes from.

tldr;

```
  def unapply(x: Any): Option[T] = x match {
    case null       => None
    case _ if runtimeClass isInstance x => Some(x.asInstanceOf[T])
    case _: Byte    if runtimeClass isAssignableFrom classOf[Byte]    => Some(x.asInstanceOf[T])
    case _: Short   if runtimeClass isAssignableFrom classOf[Short]   => Some(x.asInstanceOf[T])
    case _: Char    if runtimeClass isAssignableFrom classOf[Char]    => Some(x.asInstanceOf[T])
    case _: Int     if runtimeClass isAssignableFrom classOf[Int]     => Some(x.asInstanceOf[T])
    case _: Long    if runtimeClass isAssignableFrom classOf[Long]    => Some(x.asInstanceOf[T])
    case _: Float   if runtimeClass isAssignableFrom classOf[Float]   => Some(x.asInstanceOf[T])
    case _: Double  if runtimeClass isAssignableFrom classOf[Double]  => Some(x.asInstanceOf[T])
    case _: Boolean if runtimeClass isAssignableFrom classOf[Boolean] => Some(x.asInstanceOf[T])
    case _: Unit    if runtimeClass isAssignableFrom classOf[Unit]    => Some(x.asInstanceOf[T])
    case _ => None
```

This moves the `isInstance` case to the top, as commented, and inlines the `isAssignable` in the guards. I haven't checked if it's better to wrap the match in `Some(matched.asInstanceOf[T])` because baby steps.

Notice the pristine columnar alignment just in case this code is useful.

Here's the javap for original Moors, case _: Int, and the above.

```
        14: instanceof    #33                 // class java/lang/Byte
        17: ifeq          42
        20: aload_2      
        21: invokestatic  #90                 // Method scala/runtime/BoxesRunTime.unboxToByte:(Ljava/lang/Object;)B
        24: istore        4
        26: aload_0      
        27: iload         4
        29: invokestatic  #94                 // Method scala/runtime/BoxesRunTime.boxToByte:(B)Ljava/lang/Byte;
        32: getstatic     #37                 // Field java/lang/Byte.TYPE:Ljava/lang/Class;
        35: invokestatic  #98                 // Method scala$reflect$ClassTag$$unapplyImpl:(Lscala/reflect/ClassTag;Ljava/lang/Object;Ljava/lang/Class;)Lscala/Option;
        38: astore_3     
        39: goto          300


        14: instanceof    #33                 // class java/lang/Byte
        17: ifeq          32
        20: aload_0      
        21: aload_1      
        22: getstatic     #37                 // Field java/lang/Byte.TYPE:Ljava/lang/Class;
        25: invokestatic  #88                 // Method scala$reflect$ClassTag$$unapplyImpl:(Lscala/reflect/ClassTag;Ljava/lang/Object;Ljava/lang/Class;)Lscala/Option;
        28: astore_3     
        29: goto          213

        14: invokeinterface #19,  1           // InterfaceMethod scala/reflect/ClassTag.runtimeClass:()Ljava/lang/Class;
        19: aload_1      
        20: invokevirtual #87                 // Method java/lang/Class.isInstance:(Ljava/lang/Object;)Z
        23: ifeq          38
        26: new           #89                 // class scala/Some
        29: dup          
        30: aload_1      
        31: invokespecial #93                 // Method scala/Some."<init>":(Ljava/lang/Object;)V
        34: astore_3     
        35: goto          348
        38: aload_2      
        39: instanceof    #33                 // class java/lang/Byte
        42: ifeq          72
        45: aload_0      
        46: invokeinterface #19,  1           // InterfaceMethod scala/reflect/ClassTag.runtimeClass:()Ljava/lang/Class;
        51: getstatic     #37                 // Field java/lang/Byte.TYPE:Ljava/lang/Class;
        54: invokevirtual #97                 // Method java/lang/Class.isAssignableFrom:(Ljava/lang/Class;)Z
        57: ifeq          72
        60: new           #89                 // class scala/Some
        63: dup          
        64: aload_1      
        65: invokespecial #93                 // Method scala/Some."<init>":(Ljava/lang/Object;)V
        68: astore_3     
        69: goto          348
```

som-snytt commented:
So instead of tldr; I should have said tlacr; for too long ago, can't remember.

My comment on the previous PR meant, about whether the match does the right thing, meant, if the `instanceof Byte` fails, does it fall thru directly to the None case? And not check every case.

I'll go check on that.

Edit: the pattern match proceeds sequentially through all the type tests, unfortunately.

So as in the snippet from the PPR (previous PR), the if/else on the RHS is required for _peak_ efficiency.

Pardon, I meant :mount_fuji: efficiency.

adriaanm commented:
I couldn't come up with any value for `b`, where `b: Byte` and (thus) `b != null`, but for which `java.lang.Byte.valueOf(((java.lang.Byte)b).byteValue()) != b`. I think the same holds for Double/Float, and, erm, Unit.

adriaanm committed:
Use if/then/else to avoid box(unbox(_))

adriaanm committed:
Correct dependency for test.junit.comp

It needs a packed build. Skip the optimization of only
compiling junit tests when those source change,
since we should also rebuild when e.g., library source changes

adriaanm commented:
re-re-review by @som-snytt, please

sjrd commented:
I don't understand what the last commit has to do with this PR, but otherwise LGTM.

adriaanm commented:
I know, sorry. My only defense is that I ran into the issue with junit while writing the test and didn't feel like opening a new PR with only xml diffs. I'm kind of over xml. And ant.

som-snytt commented:
I thought Jason said ant.xml was plan B for the build.

LGTM2. I'm pretty sure this is the emoji for "nice alignment". :metal:
